File: 2020/rs/src/bin/day_09.rs
use std::fs;

fn invalid_index(nums: &[i64]) -> Option<i64> {
    for (i, &n) in nums.iter().enumerate().skip(25) {
        if !nums[i-25..i].iter().any(|&a| nums[i-25..i].contains(&(n - a))) {
            return Some(n);
        }
    }
    None
}

fn main() {
    let input = fs::read_to_string("../input/input_09.txt").unwrap();
    let nums: Vec<i64> = input.split_terminator('\n').map(|s| s.parse().unwrap()).collect();
    let invalid = invalid_index(&nums).unwrap();
    println!("Part 1: {}", invalid);
    let (mut lower, mut upper) = (0, 0);
    let mut total = nums[0];
    while total != invalid {
        if total < invalid {
            upper += 1;
            total += nums[upper]
        } else {
            total -= nums[lower];
            lower += 1;
        }
    }
    let min = nums[lower..upper+1].iter().min().unwrap();
    let max = nums[lower..upper+1].iter().max().unwrap();
    println!("Part 2: {}", min + max);
}

File: python/2020/day_09.py
import itertools as it
from pathlib import Path


def main():
    text = (Path(__file__).parent / "../input/input_09.txt").read_text()
    nums = [int(n) for n in text.splitlines()]
    for i, n in enumerate(nums):
        if i < 25:
            continue
        if not any(n == a + b for (a, b) in it.combinations(nums[i-25:i], 2)):
            print("Part 1:", invalid := n)
            break
    lower = upper = 0
    total = nums[0]
    while total != invalid:
        if total < invalid:
            upper += 1
            total += nums[upper]
        else:
            total -= nums[lower]
            lower += 1
    print("Part 2:", min(nums[lower:upper+1]) + max(nums[lower:upper+1]))


if __name__ == "__main__":
    main()
